### PR TITLE
[RELAY][GRAD] handle Tuple/TupleGetItem in first order gradient

### PR DESCRIPTION
**Pull request by t-vi**
This adds tuple support for first order gradients.

We may have tuples passed to or returned from functions and at this stage, they aren't inlined (because we want to compute gradients based on that granularity), limiting the possibility to eliminate Tuple/GetTupleItem.
Examples of ops with tupels include dropout and concatenate.

### CODE REVIEW

**Comment by t-vi** [comment]
@junrushao1994 @MarisaKirisame @tqchen for the friends of gradients...

I have a third thing (fixing an efficiency issue I found), but it also is in `gradient.cpp`, so I'll send that afterwards.

**Comment by tqchen** [comment]
also cc @yzhliu @icemelon9

File: src/relay/transforms/gradient.cc [hunk 1]
```diff
@@ -106,14 +106,67 @@ struct ADValueNode {
...
+  if (auto* tt = t.as<TensorTypeNode>()) {
+    return Ones(tt->shape, tt->dtype);
+  } else if (auto* tt = t.as<TupleTypeNode>()) {
+    std::vector<Expr> res;
+    for (size_t i = 0; i < tt->fields.size(); i++) {
+      res.push_back(MultiOnesType(tt->fields[i]));
+    }
+    return Tuple(res);
+  } else {
+    LOG(FATAL) << "unsupported type to zero: " << tt;
```
Comment (MarisaKirisame) [inline src/relay/transforms/gradient.cc:145]:
```suggestion
    LOG(FATAL) << "unsupported type to ones: " << tt;
```

File: src/relay/transforms/gradient.cc [hunk 2]
```diff
@@ -106,14 +106,67 @@ struct ADValueNode {
   }
 };
 
+Expr MultiZerosType(const Type& t) {
```
Comment (MarisaKirisame) [inline src/relay/transforms/gradient.cc:109]: can you unify MultiZeros with MultiOnes?

File: src/relay/transforms/gradient.cc (hunk 1 above)
Comment (t-vi) [reply src/relay/transforms/gradient.cc:145]: This goes away with the unification...

File: src/relay/transforms/gradient.cc (hunk 2 above)
Comment (t-vi) [reply src/relay/transforms/gradient.cc]: I just pushed that. Thank you for the suggestion.

**Pull request merged**